tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogicTest.java

```diff
@@ -547,6 +563,74 @@ public void testWholeWordSearchAvailable() {
+		expectStatusEmpty(findReplaceLogic);
+
+		findReplaceLogic.performSelectAndReplace(LINE_STRING, "");
+		assertThat(textViewer.getTextWidget().getText(), is(LINE_STRING + lineSeparator() + lineSeparator()));
+		expectStatusEmpty(findReplaceLogic);
+
+		findReplaceLogic.performSelectAndReplace(LINE_STRING, "");
+		assertThat(textViewer.getTextWidget().getText(), is(LINE_STRING + lineSeparator() + lineSeparator()));
+		expectStatusIsCode(findReplaceLogic, StatusCode.NO_MATCH);
+	}
+
+	@Test
+	public void testSearchInScopeBeginsSearchInScope() {
+		int lineSeparatorLength= lineSeparator().length();
+		TextViewer textViewer= setupTextViewer(LINE_STRING + lineSeparator() + LINE_STRING + lineSeparator() + LINE_STRING);
+		textViewer.setSelectedRange(LINE_STRING_LENGTH + lineSeparatorLength, 2 * LINE_STRING_LENGTH + lineSeparatorLength);
+		IFindReplaceLogic findReplaceLogic= setupFindReplaceLogicObject(textViewer);
+		findReplaceLogic.activate(SearchOptions.FORWARD);
+		findReplaceLogic.deactivate(SearchOptions.GLOBAL);
+		findReplaceLogic.performSearch(LINE_STRING);
+		expectStatusEmpty(findReplaceLogic);
+		assertThat(findReplaceLogic.getTarget().getSelection().x, not(is(0)));
+		assertThat(findReplaceLogic.getTarget().getSelection().x, not(is(textViewer.getDocument().get().length() - LINE_STRING_LENGTH)));
+	}
+
+
+	/**
+	 * The TextViewer implementation of IFindReplaceLogic is misleading and not adhering to the
+	 * Interface: IFindReplaceTargetExtension3#replaceSelection will NOT replace the selection if
+	 * nothing was previously found. This does not generally have to be the case and thus we mock a
+	 * target that is setup like this:
+	 *
+	 * The text contained is {@code ~SELECTEDTEXT~abcd} with ~ marking the boundaries of the
+	 * selection. We perform a search for "NOTFOUND" first - the selection stays put since the
+	 * string was not found. At this point, we do not want to perform
+	 * {@code replaceSelection("NOTFOUND")} since an implementation adhering to the specification of
+	 * the function would just overwrite the current selection.
+	 */
+	@Test
+	public void onlySelectAndReplacesIfFindSuccessfulOnCustomTarget() {
+		IFindReplaceTarget mockedTarget= Mockito.mock(IFindReplaceTarget.class, withSettings().extraInterfaces(IFindReplaceTargetExtension3.class, IFindReplaceTargetExtension2.class));
+
+		when(mockedTarget.getSelectionText()).thenReturn("SELECTEDTEXT");
+		when(mockedTarget.getSelection()).thenReturn(new Point(0, "SELECTEDTEXT".length()));
+		when(mockedTarget.isEditable()).thenReturn(true);
+		when(mockedTarget.findAndSelect(anyInt(), anyString(), anyBoolean(), anyBoolean(), anyBoolean())).thenReturn(-1);
+		when(((IFindReplaceTargetExtension2) mockedTarget).validateTargetState()).thenReturn(true);
+		doThrow(new RuntimeException("The find-string was not found so replacement is invalid at this step"))
+				.when((IFindReplaceTargetExtension3) mockedTarget)
+				.replaceSelection(anyString(), anyBoolean());
```

Comment:
https://stackoverflow.com/questions/9841623/mockito-how-to-verify-method-was-called-on-an-object-created-within-a-method